aten/src/ATen/native/TensorAdvancedIndexing.cpp

```diff
@@ -715,6 +716,75 @@ Tensor index_select_backward(const Tensor& grad, IntArrayRef self_sizes, int64_t
   return at::zeros(self_sizes, grad.options()).index_add_(dim, index, grad);
 }
 
+Tensor & index_fill_(Tensor & self, int64_t dim, const Tensor & index, Scalar source) {
+  at::NoNamesGuard guard;
+
+  TORCH_CHECK_INDEX(
+    index.scalar_type() == ScalarType::Long,
+    "index_fill_(): Expected dtype int64 for index.");
+
+  at::assert_no_overlap(self, index);
+  if (at::has_internal_overlap(self) == at::MemOverlap::YES) {
+    TORCH_WARN(
+      "Use of index_fill_ on expanded tensors is deprecated. "
+      "Please clone() the tensor before performing this operation. "
+      "This also applies to advanced indexing e.g. tensor[mask] = scalar");
+  }
+
```

Comment:
I think we should add a `TORCH_CHECK` to throw an error if we try to fill a complex value in a non-complex tensor or if we try to fill a floating point scalar in an int tensor. The latter would be a bc breaking change though:
```
>>> x = torch.tensor([[1, 2, 3], [4, 5, 6], [7, 8, 9]], dtype=torch.long)
>>> index = torch.tensor([0, 2])
>>> x.index_fill_(1, index, -1.4)
tensor([[-1,  2, -1],
        [-1,  5, -1],
        [-1,  8, -1]])
```
cc. @mruberry

Reply:
also 
```
>>> x.index_fill(1, index, -1.4)
tensor([[-1,  2, -1],
        [-1,  5, -1],
        [-1,  8, -1]])
```
I expected this to return a floating point tensor ...

Reply:
@anjali411 , you use the same `x` as from above? That one is of type `Long`. The scalar gets cast to an OpaqueType of lenght `sizeof(scalar_t)`, for `scalar_t == long` in this case.

Reply:
`index_fill_` is an inplace operation, if called on an integral tensor it cannot return floating point tensor. Out-of-place variants, with rare exceptions, should have the same behavior as inplace. For the record, indexing assignment in numpy silently converts floating point to int, but errors out on complex to int or complex to float.

Reply:
Shall we adapt the numpy's behavior regarding complex fill value?

Reply:
yeah I think we should!

Reply:
How do I get the dtype from a Scalar then?

Reply:
iirc there's `is_complex` function for Scalars

Reply:
Turns out there is `isComplex` method. So good the compiler was able to give the right suggestion from `is_complex` :)